modules/objdetect/src/qrcode.cpp

```diff
@@ -1245,5 +1228,1206 @@ cv::String QRCodeDetector::detectAndDecode(InputArray in,
+    {
+        const int x = cvRound(list_lines[pnt][0] + list_lines[pnt][2] * 0.5);
+        const int y = cvRound(list_lines[pnt][1]);
+
+        // --------------- Search vertical up-lines --------------- //
+
+        test_lines.clear();
+        uint8_t future_pixel_up = 255;
+
+        for (int j = y; j < bin_barcode.rows - 1; j++)
+        {
+            next_pixel = bin_barcode.ptr<uint8_t>(j + 1)[x];
+            temp_length++;
+            if (next_pixel == future_pixel_up)
+            {
+                future_pixel_up = static_cast<uint8_t>(~future_pixel_up);
+                test_lines.push_back(temp_length);
+                temp_length = 0;
+                if (test_lines.size() == 3) { break; }
+            }
+        }
+
+          // --------------- Search vertical down-lines --------------- //
+
+          uint8_t future_pixel_down = 255;
+          for (int j = y; j >= 1; j--)
+          {
+              next_pixel = bin_barcode.ptr<uint8_t>(j - 1)[x];
+              temp_length++;
+              if (next_pixel == future_pixel_down)
+              {
+                  future_pixel_down = static_cast<uint8_t>(~future_pixel_down);
+                  test_lines.push_back(temp_length);
+                  temp_length = 0;
+                  if (test_lines.size() == 6) { break; }
+              }
+          }
+
+          // --------------- Compute vertical lines --------------- //
+
+          if (test_lines.size() == 6)
+          {
+              double length = 0.0, weight = 0.0;
+
+              for (size_t i = 0; i < test_lines.size(); i++) { length += test_lines[i]; }
+
+              CV_Assert(length > 0);
+              for (size_t i = 0; i < test_lines.size(); i++)
+              {
+                  if (i % 3 != 0) { weight += fabs((test_lines[i] / length) - 1.0/ 7.0); }
+                  else            { weight += fabs((test_lines[i] / length) - 3.0/14.0); }
+              }
+
+              if(weight < eps_horizontal)
+              {
+                  result.push_back(list_lines[pnt]);
+              }
+          }
+      }
+      if (result.size() > 2)
+      {
+          for (size_t i = 0; i < result.size(); i++)
+          {
+              point2f_result.push_back(
+                    Point2f(static_cast<float>(result[i][0] + result[i][2] * 0.5),
+                            static_cast<float>(result[i][1])));
+          }
+      }
+      return point2f_result;
+}
+
+void QRDetectMulti::fixationPoints(vector<Point2f> &local_point)
+{
+    CV_TRACE_FUNCTION();
+    double cos_angles[3], norm_triangl[3];
+
+    norm_triangl[0] = norm(local_point[1] - local_point[2]);
+    norm_triangl[1] = norm(local_point[0] - local_point[2]);
+    norm_triangl[2] = norm(local_point[1] - local_point[0]);
+
+    cos_angles[0] = (norm_triangl[1] * norm_triangl[1] + norm_triangl[2] * norm_triangl[2]
+                  -  norm_triangl[0] * norm_triangl[0]) / (2 * norm_triangl[1] * norm_triangl[2]);
```

Comment:
You might use dot product instead of law of cosines in case of input vectors:
- https://en.wikipedia.org/wiki/Dot_product